YAML "unexpected scalar" Error when loading compilation database generated from Clang -MJ

I'm running Mull 0.5.0 for Ubuntu 18.04 running on WSL1, and launched it with the -compdb-path option pointing to a compilation database json file generated by Clang. 

I'm compiling each c source-file using Clang 6.0.0-1ubuntu2, with the -MJ command for generating a json-file per object file (*.o.json). I then use Sed to stitch together all the individual json-files into a compilation database based on this [link](https://github.com/cquery-project/cquery/wiki/Compilation-database).

My current compilation-database looks like this (only one source-file currently under test):
```JSON
[
    {
        "directory": "/root/code/c/unit-test-experiments/mull_mutation_testing/build",
        "file": "../sw/src/millisec_timer_unix.c",
        "output": "/tmp/millisec_timer_unix-19be99.bc",
        "arguments": [
            "/usr/lib/llvm-6.0/bin/clang",
            "-xc",
            "../sw/src/millisec_timer_unix.c",
            "-c",
            "-Wall",
            "-Werror",
            "-fembed-bitcode=all",
            "-g",
            "-I",
            "../sw/inc",
            "-I",
            "../sw/test/fakes",
            "-I",
            "../lib/fakefunctionframework",
            "-o",
            "obj/millisec_timer_unix.o",
            "--target=x86_64-pc-linux-gnu"
        ]
    }
]
```

The command used to run Mull was:
```
mull-cxx -test-framework=GoogleTest -mutators=default -ld_search_path=/lib/x86_64-linux-gnu/ -ld_search_path=/usr/lib/x86_64-linux-gnu/ -ld_search_path=/usr/share/gdb/auto-load/usr/lib/x86_64-linux-gnu/ -exclude-path=lib -exclude-path=sw/test -exclude-path=googletest -compdb-path=./compile_commands.json ./test
```

The error given is:
```
mull-cxx -test-framework=GoogleTest -mutators=default -ld_search_path=/lib/x86_64-linux-gnu/ -ld_search_path=/usr/lib/x86_64-linux-gnu/ -ld_search_path=/usr/share/gdb/auto-load/usr/lib/x86_64-linux-gnu/ -exclude-path=lib -exclude-path=sw/test -exclude-path=googletest -compdb-path=./compile_commands.json ./test
Extracting bitcode from executable (threads: 1): 1/1. Finished in 23ms.
Loading bitcode files (threads: 4): 5/5. Finished in 209ms.
YAML:7:13: error: unexpected scalar
            "/usr/lib/llvm-6.0/bin/clang",
            ^
Can not read compilation database: ./compile_commands.json
Compiling instrumented code (threads: 4): 5/5. Finished in 12ms.
Loading dynamic libraries (threads: 1): 1/1. Finished in 1ms.
Searching tests (threads: 1): 1/1. Finished in 0ms.
Preparing original test run (threads: 1): 1/1. Finished in 248ms.
Running original tests (threads: 2): 2/2. Finished in 305ms.
Searching mutants across functions (threads: 4): 33/33. Finished in 12ms.
Applying filter: no debug info (threads: 4): 116/116. Finished in 1ms.
Applying filter: file path (threads: 4): 92/92. Finished in 11ms.
Applying filter: junk (threads: 4): 0/10In file included from ../sw/src/millisec_timer_unix.c:1:
/usr/include/time.h:29:10: fatal error: 'stddef.h' file not found
#include <stddef.h>
         ^~~~~~~~~~
Applying filter: junk (threads: 4): 10/10. Finished in 66ms.
No mutants found. Mutation score: infinitely high

Total execution time: 895ms
```

When I spoke to Alex Denisov he suggested that this might be a bug where Mull expects the commands to be fed as a single string. 

Seeing as Clang (which is the supported compiler for the project) outputs compilation databases in array-of-string format, it would be appreciated if array-of-string support was added to Mull.

--- a/lib/JunkDetection/CXX/CompilationDatabase.cpp
+++ b/lib/JunkDetection/CXX/CompilationDatabase.cpp
@@ -12,23 +12,30 @@
 
 using namespace mull;
 
-/// Skips the next word even if it is several spaces away:
-///
-/// Examples:
-///     ' aaa bbb' ->  ' aaa bbb'
-///      ^                   ^
-///
-///     '   aaa bbb' ->  '   aaa bbb'
-///      ^                       ^
-static void skipNextWord(std::istringstream &stream) {
-  int c = stream.get();
-  while (c == ' ') {
-    stream.ignore(std::numeric_limits<std::streamsize>::max(), ' ');
-    c = stream.get();
+static std::vector<std::string> filterFlags(const std::vector<std::string> &flags,
+                                            bool stripCompiler) {
+  std::vector<std::string> filteredFlags;
+  if (flags.empty()) {
+    return filteredFlags;
   }
-  if (stream.good()) {
-    stream.unget();
+  auto it = flags.begin();
+  auto end = flags.end();
+
+  if (stripCompiler) {
+    ++it;
   }
+
+  while (it != end) {
+    std::string flag = *(it++);
+    if (flag == "-c") {
+      if (it != end) {
+        ++it;
+      }
+      continue;
+    }
+    filteredFlags.push_back(flag);
+  }
+  return filteredFlags;
 }
 
 static std::vector<std::string> flagsFromString(const std::string &s) {
@@ -41,36 +48,22 @@ static std::vector<std::string> flagsFromString(const std::string &s) {
     if (value.empty()) {
       continue;
     }
-    if (value == "-c") {
-      skipNextWord(stream);
-      continue;
-    }
     flags.push_back(value);
   }
 
   return flags;
 }
 
-const std::string &commandOrArguments(const CompileCommand &command) {
+static std::vector<std::string> flagsFromCommand(const CompileCommand &command) {
   if (command.arguments.empty()) {
-    return command.command;
+    return filterFlags(flagsFromString(command.command), true);
   }
-  return command.arguments;
+  return filterFlags(command.arguments, true);
 }
 
-static std::vector<std::string>
-flagsFromCommand(const CompileCommand &command,
-                 const std::vector<std::string> &extraFlags) {
-  const std::string &input = commandOrArguments(command);
-
-  std::istringstream stream(input);
-  std::string compiler;
-  stream >> compiler;
-
-  std::string rest;
-  std::getline(stream, rest);
-
-  std::vector<std::string> flags = flagsFromString(rest);
+static std::vector<std::string> flagsFromCommand(const CompileCommand &command,
+                                                 const std::vector<std::string> &extraFlags) {
+  std::vector<std::string> flags = flagsFromCommand(command);
   for (auto &extra : extraFlags) {
     flags.push_back(extra);
   }
@@ -78,8 +71,7 @@ flagsFromCommand(const CompileCommand &command,
 }
 
 static std::map<std::string, std::vector<std::string>>
-loadDatabaseFromFile(const std::string &path,
-                     const std::vector<std::string> &extraFlags) {
+loadDatabaseFromFile(const std::string &path, const std::vector<std::string> &extraFlags) {
   std::map<std::string, std::vector<std::string>> database;
   if (path.empty()) {
     return database;
@@ -114,11 +106,11 @@ loadDatabaseFromFile(const std::string &path,
 
 CompilationDatabase::CompilationDatabase(CompilationDatabase::Path path,
                                          CompilationDatabase::Flags flags)
-    : extraFlags(flagsFromString(flags.getFlags())),
+    : extraFlags(filterFlags(flagsFromString(flags.getFlags()), false)),
       database(loadDatabaseFromFile(path.getPath(), extraFlags)) {}
 
-const std::vector<std::string> &CompilationDatabase::compilationFlagsForFile(
-    const std::string &filepath) const {
+const std::vector<std::string> &
+CompilationDatabase::compilationFlagsForFile(const std::string &filepath) const {
   if (database.empty()) {
     return extraFlags;
   }

--- a/lib/JunkDetection/CXX/CompilationDatabaseParser.h
+++ b/lib/JunkDetection/CXX/CompilationDatabaseParser.h
@@ -2,7 +2,6 @@
 
 #include <llvm/Support/YAMLParser.h>
 #include <llvm/Support/YAMLTraits.h>
-
 #include <string>
 #include <vector>
 
@@ -12,36 +11,28 @@ struct CompileCommand {
   std::string directory;
   std::string file;
   std::string output;
-  std::string arguments;
   std::string command;
+  std::vector<std::string> arguments;
 };
 
 } // namespace mull
 
+#if LLVM_VERSION_MAJOR < 5
+LLVM_YAML_IS_SEQUENCE_VECTOR(std::string)
+#endif
+LLVM_YAML_IS_SEQUENCE_VECTOR(mull::CompileCommand)
+
 namespace llvm {
 namespace yaml {
 class Input;
-
-template <> struct SequenceTraits<std::vector<mull::CompileCommand>> {
-  static size_t size(IO &io, std::vector<mull::CompileCommand> &list) {
-    return list.size();
-  }
-
-  static mull::CompileCommand &
-  element(IO &io, std::vector<mull::CompileCommand> &list, size_t index) {
-    mull::CompileCommand t;
-    list.push_back(t);
-    return list.back();
-  }
-};
 
 template <> struct MappingTraits<mull::CompileCommand> {
   static void mapping(IO &io, mull::CompileCommand &command) {
     io.mapRequired("directory", command.directory);
     io.mapRequired("file", command.file);
     io.mapOptional("output", command.output);
-    io.mapOptional("arguments", command.arguments);
     io.mapOptional("command", command.command);
+    io.mapOptional("arguments", command.arguments);
   }
 };
 } // namespace yaml

--- a/tests/JunkDetection/CompilationDatabaseTests.cpp
+++ b/tests/JunkDetection/CompilationDatabaseTests.cpp
@@ -1,8 +1,5 @@
 #include "FixturePaths.h"
 #include "mull/JunkDetection/CXX/CompilationDatabase.h"
-
-#include <clang/Basic/Version.h>
-
 #include <gtest/gtest.h>
 
 using namespace mull;

--- a/tests/fixtures/junk_detection/compdb/db_with_arguments.json
+++ b/tests/fixtures/junk_detection/compdb/db_with_arguments.json
@@ -1,7 +1,7 @@
 [
 {
   "directory": "/foo/bar",
-  "arguments": "clang++ -I foo -I bar -c foobar.cpp",
+  "arguments": [ "clang++", "-I", "foo", "-I", "bar", "-c", "foobar.cpp"],
   "file": "foobar.cpp"
 }
 ]

--- a/tests/fixtures/junk_detection/compdb/db_with_fullpath_compiler.json
+++ b/tests/fixtures/junk_detection/compdb/db_with_fullpath_compiler.json
@@ -1,7 +1,7 @@
 [
 {
   "directory": "/foo/bar",
-  "arguments": "/opt/bin/clang++ -I foo -I bar -c foobar.cpp",
+  "command": "/opt/bin/clang++ -I foo -I bar -c foobar.cpp",
   "file": "foobar.cpp",
   "output" : "foobar.exe"
 }

--- a/tests/fixtures/junk_detection/compdb/db_with_output.json
+++ b/tests/fixtures/junk_detection/compdb/db_with_output.json
@@ -1,7 +1,7 @@
 [
 {
   "directory": "/foo/bar",
-  "arguments": "clang++ -I foo -I bar -c foobar.cpp",
+  "arguments": [ "clang++", "-I", "foo", "-I", "bar", "-c", "foobar.cpp" ],
   "file": "foobar.cpp",
   "output" : "foobar.exe"
 }

